qiskit/circuit/parameter.py

```diff
@@ -88,7 +83,12 @@ def __init__(self, name: str):
             import symengine
 
             symbol = symengine.Symbol(name)
-        super().__init__(symbol_map={self: symbol}, expr=symbol)
+
+        self._symbol_expr = symbol
+        self._parameter_keys = frozenset((self._hash_key(),))
```

Comment:
Why is `frozenset` used here?

Reply:
As opposed to the tuple, or as opposed to a regular set?  For the former: it's to match the form of the hash key of `ParameterExpression`, so they'll definitely hash the same.  As to why `_parameter_keys` is a `frozenset` as opposed to the `dict_keys` or `set` it is on `main`: it's just so it can be used directly in the hash key without a conversion.  We never need to mutate it in place, so it's safe to make it immutable.

Reply:
And why does `ParameterExpression` use a frozenset instead of a tuple? It appears to me that a tuple suffices.

Reply:
The frozenset expresses that the ordering doesn't matter (`a+b` vs `b+a` probably produce different orderings in the `_parameter_symbols` dict). A tuple could also work if it were sorted before hashing.

Reply:
If we used `tuple`, we'd have to ensure that all the `Parameter`s were always in the same order in order for it to hash and compare as expected.  That's doable, but it adds sorting overhead that it's not clear has any advantage over using a hashset-based comparison that's the more natural representation of an unordered collection.

Reply:
Oh, I see, ok.